flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -485,76 +481,151 @@ public boolean isShutdown() {
 			}
 		}
+		startTriggeringCheckpoint(
+			timestamp,
+			props,
+			externalSavepointLocation,
+			isPeriodic,
+			advanceToEndOfTime,
+			onCompletionPromise);
+		return onCompletionPromise;
+	}
 
-		// next, check if all tasks that need to acknowledge the checkpoint are running.
-		// if not, abort the checkpoint
-		Map<ExecutionAttemptID, ExecutionVertex> ackTasks = new HashMap<>(tasksToWaitFor.length);
+	private void startTriggeringCheckpoint(
+		long timestamp,
+		CheckpointProperties props,
+		@Nullable String externalSavepointLocation,
+		boolean isPeriodic,
+		boolean advanceToEndOfTime,
+		CompletableFuture<CompletedCheckpoint> onCompletionPromise) {
 
-		for (ExecutionVertex ev : tasksToWaitFor) {
-			Execution ee = ev.getCurrentExecutionAttempt();
-			if (ee != null) {
-				ackTasks.put(ee.getAttemptId(), ev);
-			} else {
-				LOG.info("Checkpoint acknowledging task {} of job {} is not being executed at the moment. Aborting checkpoint.",
-						ev.getTaskNameWithSubtaskIndex(),
-						job);
-				throw new CheckpointException(CheckpointFailureReason.NOT_ALL_REQUIRED_TASKS_RUNNING);
+		try {
+			// make some eager pre-checks
+			synchronized (lock) {
+				preCheckBeforeTriggeringCheckpoint(isPeriodic, props.forceCheckpoint());
 			}
-		}
 
-		// we will actually trigger this checkpoint!
+			final Execution[] executions = getTriggerExecutions();
+			final Map<ExecutionAttemptID, ExecutionVertex> ackTasks = getAckTasks();
+
+			// we will actually trigger this checkpoint!
+			Preconditions.checkState(!isTriggering);
+			isTriggering = true;
+
+			final CompletableFuture<PendingCheckpoint> pendingCheckpointCompletableFuture =
+				initializeCheckpoint(props, externalSavepointLocation)
+					.thenApplyAsync((checkpointIdAndStorageLocation) -> createPendingCheckpoint(
+						timestamp,
+						props,
+						ackTasks,
+						isPeriodic,
+						checkpointIdAndStorageLocation.checkpointId,
+						checkpointIdAndStorageLocation.checkpointStorageLocation,
+						onCompletionPromise), timer);
+
+			pendingCheckpointCompletableFuture
+				.thenCompose(this::snapshotMasterState)
+				.whenComplete((ignored, throwable) -> {
+					final PendingCheckpoint checkpoint =
+						FutureUtils.getWithoutException(pendingCheckpointCompletableFuture);
+
+					if (throwable == null && checkpoint != null && !checkpoint.isDiscarded()) {
+						// no exception, no discarding, everything is OK
+						snapshotTaskState(
+							timestamp,
+							checkpoint.getCheckpointId(),
+							checkpoint.getCheckpointStorageLocation(),
+							props,
+							executions,
+							advanceToEndOfTime);
+						onTriggerSuccess();
+					} else {
+						// error handling must be executed in timer thread
+						timer.execute(() -> {
+							// the initialization might not be finished yet
+							if (checkpoint == null) {
+								onTriggerFailure(onCompletionPromise, throwable);
+							} else {
+								onTriggerFailure(checkpoint, throwable);
+							}
+						});
+					}
+				});
```

Comment:
I'm a bit confused which thread is actually being used to execute what here. For example isn't the whole `whenComplete` executed already in the `timer` thread, since `CheckpointCoordinator#snapshotMasterState` returns a future that will be completed from the `timer` thread? 

If so, why do we need to call `timer.execute(()`?

Regardless if I'm wrong/right about what's executed in which thread, I think this my confusion shows something has to be changed here. Maybe always explicitly use the `executor`, `timer` parameter when scheduling/chaining something, so that the executor is clearly visible? `thenCompose(...)` -> `thenComposeAsync(..., timer/executor)`? I think that would be especially helpful as it's not easily visible/understandable what's returned from `this::snapshotMasterState`. Or at least add comments? But `selfDocumentingCode > comments`.

Reply:
> For example isn't the whole whenComplete executed already in the timer thread, since CheckpointCoordinator#snapshotMasterState returns a future that will be completed from the timer thread?

Yes, all these chain calls should be executed in timer thread, except the error handling part. The exceptional future might be completed in IO thread in `initializeCheckpoint`. 

> Maybe always explicitly use the executor, timer parameter when scheduling/chaining something, so that the executor is clearly visible? thenCompose(...) -> thenComposeAsync(..., timer/executor)? 

Yes, it is definitely clearer. Actually I had implemented it like this at the beginning. But it's not efficient for most scenarios, they are already in timer thread. Especially for the final implementation (replace timer thread by main thread), it's an unnecessary burden for main thread.

I think the perfect solution here is checking the current thread. If it's already in timer (or main) thread, it would be executed directly. Otherwise it would be scheduled in timer (or main) thread. However it's not easy (or a bit hacky) to do so for now, because timer is a `ScheduledExecutor`. My plan is to implement this after replacing the timer thread by main thread, through providing a method maybe named `ComponentMainThreadExecutor#isRunningInMainThread`.

Reply:
I'm not sure if performance matter here. This is after all a rare (at best one call every couple of seconds, more realistically one call every couple of minutes), with the overhead that should be in ~1/1000_000th of a second.

I would still go with the explicit passing of the executor, but if you prefer I would be fine with adding comments what is executed where, as this is really important for thinking about code correctness.

> Yes, all these chain calls should be executed in timer thread, except the error handling part. The exceptional future might be completed in IO thread in initializeCheckpoint.

If a failure originates from `initializeCheckpoint`, from `executor` (io), then `.whenComplete(...)` part (L551:574) are executed in the `executor` thread as well? In all of the other cases those same lines are executed in the `timer` thread? This is a bit confusing. Could we make it execute always in the same thread?

Reply:
Em... It's indeed a bit confusing. I think you have convinced me. Alright, let's make it simple, forget the tiny overhead.